Data encryption on VFS level

alexelex opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
adding encryption at the virtual file system level using AES_128_CTR algorithm
...


Detailed description / Documentation draft:
to declare an encrypted disk as a wrapper to another disk, add to the configuration
```xml
<storage_configuration>
    <disks>
        <encrypted_disk_name>
            <disk>wrapped_disk_name</disk>
            <type>encrypted</type>
            <key>enter_key</key>
            <path>path_to_encrypted_disk_on_wrapped_disk</path>
        </encrypted_disk_name>
    </disks>
</storage_configuration>
```
Each file will be encrypted using the key and AES-128-CTR algorithm 

By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.

If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ClickHouse Documentation](https://github.com/ClickHouse/ClickHouse/tree/master/docs/README.md) guide first.


Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

alexelex committed:
disk encryption init

alexelex committed:
mutex loop resolved

alexelex committed:
resolve alter table issue and add relative path to DiskEncrypted that can be declared in config as a <path>

alexelex committed:
add parameters DiskEncrypted IDisk::applyNewSettings usage

alexelex committed:
inherit DiskEncrypted from DiskDecorator; override copy (use IDisk::copy instead of LocalDisk::copy), applyNewSettings, onFreeze and removeShared*

alexelex committed:
implement read and write buffers (with debug=warning logs)

alexelex committed:
encryption initialization (without any flags like use_ssl for now)

alexelex committed:
remove debug information, fix seek issue and encryption issue

alexelex committed:
remove unnecessary includes

alexelex committed:
remove unnecessary includes

alexelex committed:
fix encryption (very small files encryption bug) and replace 64-iv to 128-iv

alexelex committed:
wide parts validation with file opening via disk

alexelex committed:
refactoring

alexelex committed:
set use_ssl flag

alexelex committed:
style fixes

alexelex committed:
encryption: throw own errors; replace nonblock getrandom to 0 getrandom and add getrandom size checking

alexelex commented:
There are two issues now:
- it might be better to set `HAVE_SYS_RANDOM_H` (or whatever, to use `sys/random.h` - `getrandom`) somehow differently, because it compiles for me without this flag, but when cmake tries `check_include_file(sys/random.h) it returns 0.
- `GRND_RANDOM` flag is better for generation random IV than 0 or `GRND_NONBLOCK` (based on [stackoverflow](https://stackoverflow.com/questions/23712581/differences-between-random-and-urandom) , [stackexchange](https://unix.stackexchange.com/questions/324209/when-to-use-dev-random-vs-dev-urandom#:~:text=There%20is%20no%20difference%20between,Apple's%20iOS%20also%20uses%20Yarrow.) ), but it works very slow. And I'm not sure which is best for CH.

alexelex committed:
style fix

alexelex committed:
use static_cast

alexelex committed:
update test_disk_types with encrypted disk

alexelex committed:
add test_encrypted_disk integration test

alexelex committed:
createDir->createDirs in DiskEncrypted initialization

alexelex committed:
add disk_s3_encrypted to integration test

alexelex committed:
add optimize on encrypted disk integration test

alexelex committed:
init retest

alexelex committed:
add file encryption unit test

alexelex committed:
style fix

alexey-milovidov commented on src/Functions/FileEncryption.h:
```diff
@@ -0,0 +1,360 @@
+    String EncryptPartialBlock(const char * partial_block, size_t size, const InitVector & iv, size_t off) const
+    {
+        if (size > block_size)
+            throw DB::Exception("Expecter partial block, got block with size > block_size: size = " + std::to_string(size) + " and offset = " + std::to_string(off),
```
Expecter -> Expected

alexelex committed:
fix exception mistake

alexelex committed:
fix uint128 initialization

alexelex committed:
style fix

alexelex committed:
Merge branch 'master' into alexelex-encryption-1

alexelex committed:
Merge branch 'master' into alexelex-encryption-1

alexelex committed:
ContextConstPtr -> ContextPtr (resolve conflicts)

Commit:
Fix ReadBufferFromS3

Commit:
Fix compilation

Commit:
Use std::random_device instead of the function get_random()

Commit:
Rename ReadEncryptedBuffer -> ReadBufferFromEncryptedFile,
WriteEncryptedBuffer -> WriteBufferFromEncryptedFile,
move file FileEncryption.h to the IO folder,
rename functions accroding to our code style,
add cpp files.

Commit:
Fix ciphering of partial blocks.

Commit:
Fix unit tests.

Commit:
Fix finalize() and sync() in WriteEncryptedBuffer.

Commit:
Fix conversion to/from big endian.

Commit:
Merge branch 'master' into alexelex-encryption-1

vitlibar commented:
@alexelex Very good work, thank you!

sevirov commented:
Internal documentation ticket: DOCSUP-12128
